docs/specs/instrumentation.md

```diff
@@ -8,34 +8,142 @@ interpreted as described in [RFC 2119].
+that cannot import Vector's events.
+
+#### Error
+
+An `<Name>Error` event MUST be emitted when an error occurs.
+
+* Properties
+  * `error_code` - An error code for the failure, if applicable.
+    * SHOULD only be specified if it adds additional information beyond
+      `error_type`.
+    * The values for `error_code` for a given error event MUST be a bounded set
+      with relatively low cardinality because it will be used as a metric tag.
+      Examples would be syscall error code. Examples of values that should not
+      be used are raw error messages from `serde` as these are highly variable
+      depending on the input. Instead, these errors should be converted to an
+      error code like `invalid_json`.
+  * `error_type` - The type of error condition. MUST be one of the types listed
+    in the `error_type` enum list in the cue docs.
+  * If any of the above properties are implicit to the specific error
+    type, they MAY be omitted from being represented explicitly in the
+    event fields. However, they MUST still be included in the emitted
+    logs and metrics, as specified below, as if they were present.
+* Metrics
+  * MUST include the defined properties as tags.
+  * MUST increment `<namespace>_errors_total` metric.
+* Logs
+  * MUST log a descriptive, user friendly error message that sufficiently
+    describes the error.
+  * MUST include the defined properties as key-value pairs.
+  * MUST log a message at the `error` level.
+  * SHOULD be rate limited to 10 seconds.
+* Events
+  * MUST emit an [`EventsDropped`] event if the error results in dropping
+    events, or the error event itself MUST meet the `EventsDropped`
+    requirements.
+
+#### EventsDropped
+
+An `<Namespace>EventsDropped` event MUST be emitted when events are dropped.
+If events are dropped due to an error, then the error event should drive the
+emission of this event, meeting the below requirements.
+
+**You MUST NOT emit this event for retriable operations that can recover and
+prevent data loss. For example, a failed HTTP request that will be retried does
+not result in data loss if the retry succeeds.** 
+
+* Properties
+  * `count` - The number of events dropped
+  * `intentional` - Distinguishes if the events were dropped intentionally. For
+    example, events dropped in the `filter` transform are intentionally dropped,
+    while events dropped due to an error in the `remap` transform are
+    unintentionally dropped.
+  * `reason` - A short, user-friendly reason that describes why the events were
+    dropped.
+* Metrics
+  * MUST increment the `<namespace>_discarded_events_total` counter by the
+    number of events discarded.
+  * MUST include the listed properties as tags except the `reason` property.
+* Logs
+  * MUST log a `<count> events <intentional> dropped: <reason>` message.
+  * If `intentional` is `true`, MUST log at the `debug` level.
+  * If `intentional` is `false`, MUST log at the `error` level.
```

Comment:
Something still feels weird to me about this tag, but going with it for now. My other thought is making this an enum like:

- errored
- throttled
- sampled
- filtered
- unconsumed

Which will make it much clearer why the events were dropped.

Reply:
My only concern with an enum is maintaining the definition of intentional/unintentional reasons. The idea of `intentional` was to drive alerts in the UI. With an enum we could add a new unintentional reason that would not be recognized in the UI, etc.

Reply:
Yeah, true, I guess we can always add the enum later, in addition, and keep this flag.